File: Difficulty: Medium/Longest Consecutive Subsequence/longest-consecutive-subsequence.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            // Read first array
            String input = br.readLine();
            String[] inputArray = input.split(" ");
            int[] arr = Arrays.stream(inputArray).mapToInt(Integer::parseInt).toArray();

            Solution ob = new Solution();
            int res = ob.longestConsecutive(arr);

            System.out.println(res);
            System.out.println("~");
        }
    }
}

// } Driver Code Ends


class Solution {

    // Function to return length of longest subsequence of consecutive integers.
    public int longestConsecutive(int[] arr) {
        HashSet<Integer> st = new HashSet<>();
        int res = 0;
        for (int val : arr)
            st.add(val);
        for (int val : arr) {
            if (st.contains(val) && !st.contains(val - 1)) {
                int cur = val, cnt = 0;
                while (st.contains(cur)) {
                    st.remove(cur);
                    cur++;
                    cnt++;
                }
                res = Math.max(res, cnt);
            }
        }
        return res;
    }
}

File: Difficulty: Medium/Longest Consecutive Subsequence/longest-consecutive-subsequence.cpp
class Solution {
  public:
    int longestConsecutive(vector<int>& arr) {
        // code here
        int n=arr.size();
        sort(arr.begin(),arr.end());
        int maxLen = 1;
        int currentLen = 1;
        for(int i=0; i<n-1; i++)
        {
            int diff = arr[i+1] - arr[i];
            if(diff == 1){
                currentLen++;
                maxLen = max(maxLen, currentLen);
            }else if(diff == 0){
                continue;
            }else{
                currentLen = 1;
            }
        }
        return maxLen;
    }
    
    //1 9 3 10 4 20 2 
    //1 2 3 4 9 10 20
};